### ISSUE TITLE
[NumberField]: onValueChange not being called when input value is changed

### ISSUE THREAD

**Issue by foobarvaz** [mui/base-ui#579]
### Search keywords

onValueChange

### Latest version

- [X] I have tested the latest version

### Steps to reproduce

Link to live example:
https://codesandbox.io/p/sandbox/wpqjmg?file=%2Fsrc%2FApp.tsx%3A10%2C43-10%2C56

Steps:
1. Focus on input and change it's value

### Current behavior

`onValueChange` not being called when input value is changed

You can see that the text below the input will remain `value: 0` until you focus out of the input. The blur leads to `onValueChange` being called.

### Expected behavior

`onValueChange` **being** called when input value is changed

### Context

Thanks for adding support for decimals in the new version of the number input ❤️ 

I'm working on an interactive calculator in which changes to input values should be reflected in calculations that are shown in the UI.

I tried to debug the issue and I found that on each change I reach this part of the code:

https://github.com/mui/base-ui/blob/78d4abd6a72b0e3d783aa7a09bfda8c7a7518b88/packages/mui-base/src/NumberField/Root/useNumberFieldRoot.ts#L607

Because `event.isTrusted === true` in this case, we call only `setInputValue` without calling `setValue` (which internally calls `onValueChange`)

I'm not sure if it's by design or not. So forgive me if I classified this issues as bug 🐛  by mistake.

### Your environment

<details>
  <summary><code>npx @mui/envinfo</code></summary>

```
  Don't forget to mention which browser you used.
  Output from `npx @mui/envinfo` goes here.
```
</details>

**Comment by atomiks** [comment]
> I'm not sure if it's by design or not. So forgive me if I classified this issues as bug 🐛 by mistake.

Yes it's by design. The reason is partial input may not be parseable as a valid number, so it can't properly be called in realtime/on change.

**Comment by bengry** [comment]
> > I'm not sure if it's by design or not. So forgive me if I classified this issues as bug 🐛 by mistake.
> 
> Yes it's by design. The reason is partial input may not be parseable as a valid number, so it can't properly be called in realtime/on change.

Can this potentially be made available on the hook variant of `NumberField`? Or an optional prop that defaults to `false` (= immediate change behavior disabled).

I understand the concern, but it gives a worse UX (less immediate feedback) because of an edge case, for a lack of a better term - when a user enters invalid values. In the happy flow path, the UX is just worse without any benefit.
E.g. [`react-number-format`](https://www.npmjs.com/package/react-number-format) deals fine with this - [demo](https://stackblitz.com/edit/vitejs-vite-h85m6c?file=src%2FApp.tsx).

**Comment by atomiks** [comment]
I think the main concern is with compositional inputs such as pinyin, which depends on the client's locale/number system. The input already blocks most invalid input before it can be typed in, but composition events aren't blocked yet represent incomplete data that can't be parsed?

This behavior matches the RA library, though it doesn't state what "partial input":

> The onChange event is triggered whenever the number value updates. This happens when the user types a value and blurs the input, or when incrementing or decrementing the value. It does not happen as the user types because partial input may not be parseable to a valid number.

Open to suggestions or PRs to improve this behavior however

**Comment by bengry** [comment]
> > The onChange event is triggered whenever the number value updates. This happens when the user types a value and blurs the input, or when incrementing or decrementing the value. It does not happen as the user types because partial input may not be parseable to a valid number.
> 
> Open to suggestions or PRs to improve this behavior however

I suggest the `onValueChange` fires only if the parsed input is valid.
e.g.
<kbd>1</kbd> → `onValueChange(1)`
<kbd>2</kbd> → `onValueChange(12)`
<kbd>.</kbd> → onValueChange **not** fired
<kbd>3</kbd> → `onValueChange(12.3)`

**Comment by oliviertassinari** [comment]
It sounds like this issue is about clearly making the difference between two separate states:

1. What's in the text box input (missing today)
2. What's the value representation of 1. (called `onValueChange` today)

If we take the Material UI API naming paradigm, 1. would be `onInputChange` https://mui.com/material-ui/api/autocomplete/#autocomplete-prop-onInputChange, and 2. would be `onChange`.

In #2105, we called 1. `onInputValueChange`, we could do the same here.

---

A side note, it seems that we have different names here, connecting this to #600.

**Comment by bengry** [comment]
> 1. What's in the text box input (missing today)

This means that the component must be controlled though, no? Or at the very least - moves the burden of parsing the data to see if the number is valid (yet) to the consumer. When `onInputChange` _can_ update on every keystroke, and not just on blur, since it already handles the parsing.
LMK if I missed something though.

**Comment by atomiks** [comment]
@bengry it does seem ok to me that `onValueChange` is called as soon as it can, as long as the current number string is parseable/valid. 

> 1. What's in the text box input (missing today)

@oliviertassinari In our API, instead of `onInputValueChange` you can just use `onChange` on the `Input` subcomponent to read the string value in realtime.

**Comment by mj12albert** [comment]
> Open to suggestions or PRs to improve this behavior however

Let's label this "ready to take", if anybody is interested and has an idea of how to do this we can discuss here ~

**Comment by oliviertassinari** [comment]
> In our API, instead of onInputValueChange you can just use onChange on the Input subcomponent to read the string value in realtime.

@atomiks Right ok, this makes sense.

But then why do we have `onInputValueChange()` in https://deploy-preview-2105--base-ui.netlify.app/react/components/combobox#root?

> it does seem ok to me that onValueChange is called as soon as it can, as long as the current number string is parseable/valid.

We could challenge this, e.g., #859 onInput() vs. onChange() on a `<input type="range">`:

- `input` is for as soon as the value changes: https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/change_event
- `change` is for once the value is committed: https://developer.mozilla.org/en-US/docs/Web/API/Element/input_event

But yeah, I see no notion of "commit" for the NumberField, so today, when it waits for a blur to trigger onChange when typing but it doesn't wait when using the keyboard look clearly broken:

https://github.com/user-attachments/assets/07dad2d5-4c87-457f-83d1-fc51cd87ec13

https://codesandbox.io/p/sandbox/number-field-forked-k59zmr?file=%2Fpackage.json
